cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/common/DeterministicCluster.java

```diff
@@ -638,6 +647,961 @@ public static ClusterModel minLeaderReplicaPerBrokerSatisfiable3() {
+
+    // Create topic partitions.
+    TopicPartition topic0Partition0 = new TopicPartition(TOPIC0, 0);
+    TopicPartition topic0Partition1 = new TopicPartition(TOPIC0, 1);
+
+    TopicPartition topic1Partition0 = new TopicPartition(TOPIC1, 0);
+    TopicPartition topic1Partition1 = new TopicPartition(TOPIC1, 1);
+
+    AggregatedMetricValues aggregatedMetricValues =
+            getAggregatedMetricValues(TestConstants.TYPICAL_CPU_CAPACITY / 8,
+                                      TestConstants.LARGE_BROKER_CAPACITY / 8,
+                                      TestConstants.MEDIUM_BROKER_CAPACITY / 8,
+                                      TestConstants.LARGE_BROKER_CAPACITY / 8);
+
+    List<Long> windows = Collections.singletonList(1L);
+    // Create replicas for topics.
+    // T0_P0_leader
+    cluster.createReplica(RACK_BY_BROKER5.get(0).toString(), 0, topic0Partition0, 0, true);
+    cluster.setReplicaLoad(RACK_BY_BROKER5.get(0).toString(), 0, topic0Partition0, aggregatedMetricValues, windows);
+    // T0_P1_leader
+    cluster.createReplica(RACK_BY_BROKER5.get(2).toString(), 2, topic0Partition1, 0, true);
+    cluster.setReplicaLoad(RACK_BY_BROKER5.get(2).toString(), 2, topic0Partition1, aggregatedMetricValues, windows);
+
+    // T1_P0_leader
+    cluster.createReplica(RACK_BY_BROKER5.get(3).toString(), 3, topic1Partition0, 0, true);
+    cluster.setReplicaLoad(RACK_BY_BROKER5.get(3).toString(), 3, topic1Partition0, aggregatedMetricValues, windows);
+    // T1_P1_leader
+    cluster.createReplica(RACK_BY_BROKER5.get(5).toString(), 5, topic1Partition1, 0, true);
+    cluster.setReplicaLoad(RACK_BY_BROKER5.get(5).toString(), 5, topic1Partition1, aggregatedMetricValues, windows);
+
+    // T0_P0_follower_1
+    cluster.createReplica(RACK_BY_BROKER5.get(1).toString(), 1, topic0Partition0, 1, false);
+    cluster.setReplicaLoad(RACK_BY_BROKER5.get(1).toString(), 1, topic0Partition0, aggregatedMetricValues, windows);
+    // T0_P1_follower_1
+    cluster.createReplica(RACK_BY_BROKER5.get(0).toString(), 0, topic0Partition1, 1, false);
+    cluster.setReplicaLoad(RACK_BY_BROKER5.get(0).toString(), 0, topic0Partition1, aggregatedMetricValues, windows);
+
+    // T1_P0_follower_1
+    cluster.createReplica(RACK_BY_BROKER5.get(4).toString(), 4, topic1Partition0, 1, false);
+    cluster.setReplicaLoad(RACK_BY_BROKER5.get(4).toString(), 4, topic1Partition0, aggregatedMetricValues, windows);
+    // T1_P1_follower_1
+    cluster.createReplica(RACK_BY_BROKER5.get(6).toString(), 6, topic1Partition1, 1, false);
+    cluster.setReplicaLoad(RACK_BY_BROKER5.get(6).toString(), 6, topic1Partition1, aggregatedMetricValues, windows);
+
+    return cluster;
+  }
+
+  /**
+   * 3 Brokers in 2 BrokerSets
+   * <p>
+   * 2 Topics
+   * <p>
+   * This would be unsatisfiable since the Disk Distribution would try to swap T0_P0_leader and T1_P0_leader but BrokerSetAwareGoal
```

Comment:
Since BrokerSetAwareGoal is a hard goal and should be prioritized, I assume the DiskDistrubutionGoal should not swap the leader, and thus the BrokerSetAwareGoal should be still satisfied?

Reply:
This is just added for coverage of cases where any other goal is prioritized over BrokerSetAwareGoal. The soft goal being prioritized over hard indeed is not expected. And hence these two tests -
```
    p.add(params(balancingConstraint, DeterministicCluster.brokerSetUnSatisfiable4(),
                 Arrays.asList(DiskUsageDistributionGoal.class.getName(),
                               BrokerSetAwareGoal.class.getName()), verifications, OptimizationFailureException.class));
    // Since DiskUsageDistribution is Soft Goal, the exception won't be thrown if the optimization suggested does not abide by BrokerSetAwareGoal
    p.add(params(balancingConstraint, DeterministicCluster.brokerSetUnSatisfiable4(),
                 Arrays.asList(BrokerSetAwareGoal.class.getName(),
                               DiskUsageDistributionGoal.class.getName()), verifications, null));

```

Reply:
Can we resolve this if explanation makes sense ?